### PR TITLE
Add a function for searching UniProt

### PR DESCRIPTION
**Pull request by JureFabjan**
The added function performes a search on UniProt and returns the hits in aspecified format.

This pull request addresses issue #...

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

- [x] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by peterjc** [comment]
This looks reasonable, although it would need some tests and perhaps inclusion in the Tutorial somewhere to help people find it and start using it.

e.g. to start with, add something to ``test_SeqIO_online.py``, and then perhaps look at the style of ``test_TogoWS.py`` for a more in depth set of tests.

Does the UnitProt API documentation talk about the encoding they use? Right now the code will assume the local system's default encoding when turning the bytes received into (unicode) strings.

**Comment by JureFabjan** [comment]
The adding of tests is no problem. I will do it when I resolve the points I bring up in this comment. And of course I will gladly also contribute to the tutorial.

The UniPort API documentation does not talk specifically about the encoding they use. As far as I tested it, the latin-1 encoding works great where the encoding is not pre-specified (where it is utf-8). The further change I would make to the function is in the type of object it returns. The user can specify the format as one of the following: html | tab | xls | fasta | gff | txt | xml | rdf | list | rss. In case of fasta, for example, it would make more sense, if the function returned the results as the Bio.SeqIO.FastaIO.SimpleFastaParser object.

**Comment by peterjc** [comment]
If the data is returned as FASTA, then we still want to get it as a handle. This is the most flexible API, and is consistent with other online resources in Biopython like TogoWS and the NCBI Efetch.

The user can then easily give that handle, to ``SeqIO.parse(...)`` to get ``SeqRecord`` objects, ``Bio.SeqIO.FastaIO.SimpleFastaParser(...)`` to get tuples of strings, write it to a file, or anything else.

**Comment by JureFabjan** [comment]
I see. Then I will not change it anymore. I have added the test in the correct file and created a new pull request for that.

**Comment by peterjc** [comment]
Please include the new test here (and not on a separate pull request like #1763 because the test is not useful on its own - it is part of this enhancement).

File: Tests/test_SeqIO_online.py [hunk 1]
```diff
@@ -95,6 +96,19 @@ def funct(d, f, e, l, c):
     del funct
 del database, formats, entry, length, checksum
 
+class UniProtTests(unittest.TestCase):
+    """Test for UniPort module."""
+
+    def test_search(self):
```
Comment (peterjc) [inline Tests/test_SeqIO_online.py]: This generic UniProt search test would be better in a new file I think, say ``test_UniProt_online.py``.

This file (``test_SeqIO_online.py``) was intended for parsing various sequence files from online resources with Bio.SeqIO, as a test of the parser (e.g. has the file format changed in a new release) and of the online wrapper (has the API changed, or is it offline). Could you use the UniProt FASTA return mode here, and any other sequence formats ``SeqIO`` can parse?

Comment (peterjc) [reply Tests/test_SeqIO_online.py]: That looks good, thanks.

File: Tests/test_SeqIO_online.py [hunk 2]
```diff
@@ -95,6 +96,26 @@ def funct(d, f, e, l, c):
     del funct
 del database, formats, entry, length, checksum
 
+class UniProtTests(unittest.TestCase):
+    """Test for UniPort module."""
```
Comment (peterjc) [inline Tests/test_SeqIO_online.py]: Typo, and maybe say UniProt search here?

**Comment by JureFabjan** [comment]
Indeed. I will update it and also create a separate test file for testing if the function itself works.

File: Bio/UniProt/__init__.py [hunk 3]
```diff
@@ -11,3 +11,33 @@
...
+from Bio._py3k import urlopen as _urlopen
+from Bio._py3k import urlencode as _urlencode
+from Bio._py3k import _binary_to_string_handle
+
+
+def search(text, output_format="tab", sort="score", oragnism="", columns=(),
+           isoform=False, compress=False, offset=0, limit=0):
+    """Perform a query over the UniProt API.
+
+    More at: https://www.uniprot.org/help/api_queries
```
Comment (peterjc) [inline Bio/UniProt/__init__.py]: We should probably provide a little more help here, specifically that columns should be a list/tuple of strings, and perhaps that isoform and compress are treated as booleans, with offset and limit expected to be integers.

**Comment by peterjc** [comment]
The TravisCI style checks are failing,

```
$ flake8 Tests/
Tests/test_SeqIO_online.py:99:1: E302 expected 2 blank lines, found 1
Tests/test_SeqIO_online.py:112:18: F821 undefined name 'search'
Tests/test_SeqIO_online.py:119:1: E305 expected 2 blank lines after class or function definition, found 1
```

This is easiest to fix locally if you have ``flake8`` installed, see https://github.com/biopython/biopython/blob/master/CONTRIBUTING.rst

**Comment by JureFabjan** [comment]
Sorry for a bit of a long pause in adapting the pull request.
Now I added a piece of documentation explaining the possibilities for columns and output format (as this also has limitations on the formats, which can be chosen.

**Comment by rwst** [comment]
I would try to fix the codecov fails myself but AFAIK this works only by submitting a PR. @JureFabjan do you agree with me opening a new PR with your commits?

**Comment by peterjc** [comment]
We're not enforcing the CodeCov metrics, but they are a useful guide to when changes need more tests.

**Comment by peterjc** [comment]
There is no merge conflict here (currently), the only issue I can see is the typo UniPort

**Comment by JureFabjan** [comment]
@peterjc do I need to change anything else, or can this branch be merged?

**Pull request closed without merging**